source/common/router/router.cc

```diff
@@ -1040,6 +1042,7 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
     if (encode_complete_) {
       stream_info_.onLastUpstreamTxByteSent();
       parent_.callbacks_->streamInfo().onLastUpstreamTxByteSent();
+      parent_.callbacks_->upstreamRequestComplete();
```

Comment:
Just going to point out that this seems like it may be called multiple times (because encodeData and encodeTrailers might call it).

Reply:
FWIW, I'm uncomfortable with upstreamRequestComplete() being called multiple times from an API definition perspective.  Having said that, I have the same concern about onLastUpstreamTxByteSent(), and it seems like the concern is the same there?  It looks like for this to be called multiple times would require encodeTrailer() to be called after encode_complete_ was set, which I'd think would be a contract violation--am I missing something about the flow in this class?